### PR TITLE
Send/Receive Encoded Content

### PR DESCRIPTION
**Pull request by daria-github**
This PR adds the ability to send and receive encoded content on iOS and Android.
Android -
Screenshot 2023-06-29 at 5 52 24 PM

iOS -

### CODE REVIEW

**Comment by nplasterer** [comment]
I don't think you can do the kotlin modules in the next PR. It needs to be done in this one. This is a breaking change. Sending messages won't work at all in android now because you removed the `sendMessage` function. Since other people are cutting npm releases in this repo you can't guarantee that a release won't get cut with this change making the android experience entirely broken.

**Comment by daria-github** [comment]
Makes sense, will do this in 1 go — just updated the overview. Thanks!

File: example/src/ConversationView.tsx [hunk 1]
```diff
@@ -8,25 +9,44 @@ import {
...
+import { TextCodec } from "./test_utils";
+import {
+  CodecRegistry,
+  ContentCodecInterface,
+} from "../../src/lib/CodecRegistry";
 
 type Props = NativeStackScreenProps<RootStackParamList, "Conversation View">;
 
+type DecodedMessage = {
+  content: string;
```
Comment (nakajima) [inline example/src/ConversationView.tsx]: I think this should be `any` now that content can be whatever there's a content codec for?
```suggestion
  content: any;
```

File: example/src/test_utils.ts [hunk 2]
```diff
@@ -1,41 +1,85 @@
...
       versionMinor: 1,
     });
   }
 
-  encode(content: Uint8Array) {
+  encode(content: string) {
     const encodedContent = {
       type: this.contentType,
       content: Buffer.from(JSON.stringify(content)),
       fallback: "fallbackText",
```
Comment (nakajima) [inline example/src/test_utils.ts]: I think the fallback for text should just be the content string itself?

File: ios/Wrappers/BinaryDataWrapper.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+enum BinaryDataWrapperError: Swift.Error {
+	case encodeError(String)
+}
+
+protocol BinaryDataWrapper: Codable {
+	associatedtype T
+	static func wrap(model: T) throws -> Self
+}
+
+extension BinaryDataWrapper {
```
Comment (nakajima) [inline ios/Wrappers/BinaryDataWrapper.swift:18]: Not an issue for this PR, but I'm curious how this performs vs the JSON wrapper we're using for the other types. If it's faster we might want to move the other wrappers to use this as well. Just an idle thought, no action needed here.

File: ios/Wrappers/BinaryDataWrapper.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  static func encode(_ model: T) throws -> [UInt8] {
+    let wrapper = try wrap(model: model)
+	let encodedData = try MessagePackEncoder().encode(wrapper)
```
Comment (nakajima) [inline ios/Wrappers/BinaryDataWrapper.swift:21]: Looks like we've got some tabs vs spaces issues in this file.

File: ios/Wrappers/EncodedMessageWrapper.swift [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import Foundation
+import XMTP
+
+// Wrapper around XMTP.EncodedMessage to allow passing these objects back
+// into react native.
+struct EncodedMessageWrapper: BinaryDataWrapper {
+	static func wrap(model: XMTP.DecodedMessage) throws -> EncodedMessageWrapper {
+		return EncodedMessageWrapper(
+			id: model.id,
+            content: [UInt8](try model.encodedContent.serializedData()),
```
Comment (nakajima) [inline ios/Wrappers/EncodedMessageWrapper.swift]: Looks like more tabs vs spaces issues.

File: src/index.ts [hunk 6]
```diff
@@ -58,18 +64,26 @@ export async function listMessages(
...
+    return decode(message);
   });
+
+  for (const message of mappedMessages) {
+    const encodedContent = proto.content.EncodedContent.decode(message.content);
+
+    message.content = encodedContent;
+  }
+
+  return mappedMessages;
```
Comment (nakajima) [inline src/index.ts]: Maybe we should move this into the `map` above so we don't need to loop over the message list twice?

**Review by nakajima — APPROVED** [review]
Left a couple small questions/comments, I think we can punt on the formatting for a follow up PR though to keep this one focused.

**Review by nplasterer — APPROVED** [review]
LGTM! Great work on this. I did a quick lint on the android files (removing unused imports and reformatting) and just pushed that in a commit.

**Pull request merged**